File: 03월 25일 - 백트래킹/과제/15663.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

const int SIZE = 8; //배열 크기

int n, m;
vector<int> num(SIZE);
vector<int> ans(SIZE);
vector<bool> check(SIZE);

void backtracking(int cnt) {
    if (cnt == m) { //길이 m 이라면 (기저조건)
        for (int i = 0; i < m; i++) {
            cout << ans[i] << ' ';
        }
        cout << '\n';
        return;
    }
    int prev = 0; //이전에 선택한 값
    for (int i = 0; i < n; i++) {
        if (!check[i] && prev != num[i]) { //수 위치 중복 체크 && 이전에 탐색한 값이 아니라면
            prev = num[i];
            ans[cnt] = num[i];
            check[i] = true;
            backtracking(cnt + 1);
            check[i] = false;
        }
    }
}

/**
 * 1. 우선 수열을 사전 순으로 증가하는 순서로 출력해야 하므로 주어진 수열을 오름차순 정렬
 * 2. 이 상태에서 우선, 같은 위치의 수를 또 선택하지 않도록 중복제거 (check 배열 사용)를 해줌
 * 3. 그 후, 중복되는 "수열"을 거르는 방법은 이전에 선택한 값을 변수에 저장해서, 수열을 만들다 다시 루트 노드로 돌아왔을 때
 *    이전에 선택한 값과 같은 값이면 넘어가면 됨! -> 어차피 같은 수이므로 같은 과정 반복해서 똑같은 수열이 나올 것
 */

int main() {
    //입력
    cin >> n >> m;
    for (int i = 0; i < n; i++) {
        cin >> num[i];
    }

    //연산
    sort(num.begin(), num.begin() + n);

    //백트래킹 & 출력
    backtracking(0);
    return 0;
}

File: 03월 25일 - 백트래킹/과제/15663.py
import sys
input = sys.stdin.readline


"""
 1. 우선 수열을 사전 순으로 증가하는 순서로 출력해야 하므로 주어진 수열을 오름차순 정렬
 2. 이 상태에서 우선, 같은 위치의 수를 또 선택하지 않도록 중복제거 (check 배열 사용)를 해줌
 3. 그 후, 중복되는 "수열"을 거르는 방법은 이전에 선택한 값을 변수에 저장해서, 수열을 만들다 다시 루트 노드로 돌아왔을 때
    이전에 선택한 값과 같은 값이면 넘어가면 됨! -> 어차피 같은 수이므로 같은 과정 반복해서 똑같은 수열이 나올 것
"""

def backtracking(idx, m):
    if idx == m:
        print(*answer) # *list -> 리스트의 요소를 하나씩 풀어서 print()에 인자로 넣어줌
                       # print(*[1, 2, 3]) == print(1, 2, 3)
        return

    prev = 0    # 이전에 선택한 값
    for i in range(n):
        if not checked[i] and arr[i] != prev:
            checked[i] = True
            prev = arr[i]
            answer[idx] = arr[i]
            backtracking(idx + 1, m)
            checked[i] = False

    return
            
n, m = map(int, input().split())
arr = list(map(int, input().split()))
arr.sort()
checked = [False] * n
answer = [0] * m

backtracking(0, m)